File: 015.go
/*
Starting in the top left corner of a 2×2 grid, and only being able to move to the right and down, there are exactly 6 routes to the bottom right corner.

How many such routes are there through a 20×20 grid?

(2n)!/(n!)^2
*/

package main

import "fmt"

func main() {
	fmt.Println(latticeRoutes(20))
}

func latticeRoutes(n int) int {
	dividend := factorialTop(n)
	divisor := factorialArr(n)
	fmt.Println(dividend, " / ", divisor)

	// need to simplify fraction in order to
	// fit numbers
	simplify(dividend, divisor)
	fmt.Println(dividend, " / ", divisor)
	product1 := 1
	for _, i := range dividend {
		product1 *= i
	}
	product2 := 1
	for _, i := range divisor {
		product2 *= i
	}
	return product1 / product2
}

func simplify(dividend []int, divisor []int) {
	for i, _ := range dividend {
		for j, _ := range divisor {
			if (divisor[j] != 1) && (dividend[i]%divisor[j]) == 0 {
				fmt.Println("dividing:", dividend[i], divisor[j])
				dividend[i] = dividend[i] / divisor[j]
				divisor[j] = 1
			}
		}
	}
}

func factorialTop(n int) []int {
	var ret []int
	for i := (n + 1); i <= (n * 2); i++ {
		ret = append(ret, i)
	}
	return ret
}
func factorialArr(n int) []int {
	var ret []int
	for i := 1; i <= n; i++ {
		ret = append(ret, i)
	}
	return ret
}

/*
func factorial(n int) int {
	product := 1
	for i := 1; i <= n; i++ {
		product *= i
	}
	return product
}
*/

File: 015.rs
fn main() {
    println!("{}", lattice_routes(20));
}

fn lattice_routes(n: usize) -> usize {
    let mut dividend = fact_dividend(n);
    let mut divisor = fact_divisor(n);
    simplify_factors(&mut dividend, &mut divisor);
    return calculate(dividend, divisor);
}

// calculate products of factors in a fraction
fn calculate(dividend: Vec<usize>, divisor: Vec<usize>) -> usize {
    let mut product_dividend: usize = 1;
    let mut product_divisor: usize = 1;
    for i in dividend.iter() {
        product_dividend *= i;
    }
    for i in divisor.iter() {
        product_divisor *= i;
    }
    return (product_dividend / product_divisor);
}

// simplify a fraction in place
fn simplify_factors(dividend: &mut Vec<usize>, divisor: &mut Vec<usize>) {
    for i in 0..dividend.len() {
        for j in 0..divisor.len() {
            if (divisor[j] != 1) && (dividend[i] % divisor[j] == 0) {
                //println!("Reducing {} / {} -> {} / 1", dividend[i], divisor[j], dividend[i] / divisor[j]);
                dividend[i] = dividend[i] / divisor[j];
                divisor[j] = 1;
            }
        }
    }
}

// produce a vector of factors for (2n)!
fn fact_dividend(n: usize) -> Vec<usize> {
    let mut ret: Vec<usize> = Vec::new();
    for i in (n+1)..=(n*2) {
        ret.push(i);
    }
    return ret;
}

// produce a vector of factors for (n!)^2
fn fact_divisor(n: usize) -> Vec<usize> {
    let mut ret: Vec<usize> = Vec::new();
    for i in 1..=n {
        ret.push(i);
    }
    return ret;
}
